### ISSUE TITLE
Make TaskReportContainer public to ease implementation of ReportContainer

### ISSUE THREAD

**Issue by 3flex** [gradle/gradle#7063]
### Expected Behavior
3rd party code quality plugins can implement `TaskReportContainer`.

### Current Behavior
3rd party code quality plugins cannot implement `TaskReportContainer` as it's part of the internal package namespace.

### Context
The docs for [`Reporting`](https://docs.gradle.org/current/dsl/org.gradle.api.reporting.Reporting.html) say to implement the reporting interface just do the following:

```
class MyTask implements Reporting<MyReportContainer> {
    // implementation
}
```

But to implement the interface an implementation for approximately 40 methods is required.

Gradle's code quality plugins can take advantage of internal class `TaskReportContainer` which simplifies things greatly, but as it's part of the internal API it's not available to well-behaved 3rd party plugins.

### Steps to Reproduce (for bugs)
N/A

### Your Environment
Trying to implement `ReportContainer` in [detekt](https://github.com/arturbosch/detekt).

**Comment by adammurdoch** [comment]
The `Reporting` interface isn't really intended to be implemented outside of the Gradle core plugins. We certainly could make it possible to implement it using only public types or methods. We probably wouldn't do this by making the `TaskReportContainer` type public but might instead provide factory methods for creating `ReportContainer` instances.

However, I think it might be better to reconsider the report APIs so that the reports generated by a task are described using the same pattern that we use to describe other kinds of inputs and outputs, rather than this entirely different pattern. For example, introducing an @OutputReport annotation.

It might be better to change the title of this issue to something like "make it possible for tasks and plugins to generate custom reports".

**Comment by ldaley** [comment]
Tasks can already generate reports. The only thing that `Reporting` really provides is a semi-consistent mechanism for users to configure report generation for a task. There are potentially other useful things that Gradle could do with the knowledge of where a task produced its reports, but it doesn't today.

The title should probably be “make it possible for tasks to allow configuring their reporting in a consistent manner” or some such.

**Comment by ldaley** [comment]
@3flex apologies, I was more responding to @adammurdoch's comments.

> If Reporting is only intended to be implemented by the core plugins, that means the Build Dashboard is intended for core plugins only as well?

Effectively, yes. I would be surprised if anyone is seriously using the build dashboard plugin.

I agree with you that the current situation is unfortunate and unkind. In your case, what is motivating you to use `Reporting`? Anything in particular? Or are you just trying to follow the lead of the core plugins?

**Comment by 3flex** [comment]
> I would be surprised if anyone is seriously using the build dashboard plugin.

I agree, just pointing out that the docs there also make it seem like a 3rd party code quality plugin would want to implement `Reporting` to show up in the dashboard, because why wouldn't other tools want closer integration with built-in Gradle features? (not being rude, just stating my thought process - reading things like that in the docs provide motivation for me to "do the right thing", because who knows which other features in future might also require `Reporting` to be implemented).

> trying to follow the lead of the core plugins

Exactly this - guided by what looked like a recommendation from the docs.  Seems it shouldn't have been too hard to have the custom task implement `ReportContainer` and expose config for a single `xml` and `html` report which only required customisation of their destinations based on [this example](https://docs.gradle.org/current/dsl/org.gradle.api.reporting.Reporting.html).

Simplest thing for now would be to update the docs and remove references to implementations that aren't feasible, and possibly also state in docs that certain interfaces are documented publically but only intended to be implemented internally. Request to make `TaskReportContainer` public was a moonshot :)

A factory to create a `ReportContainer` would be welcome but I doubt it's valuable enough to justify building and maintaining.

**Comment by ldaley** [comment]
Thanks for clarifying.

Part of the problem here is that `Reporting` doesn't work very well. It's cumbersome to implement and is awkward for a bunch of uses. If it was really good, we'd be much more enthusiastic about opening it up because there is no good reason why explicit modelling of reports should be confined to core tasks and good reasons for having a comprehensible pattern here.

I'll take a look at the docs. In the meantime, I suggest modelling your reporting however best makes sense for your task.

**Comment by ldaley** [comment]
@oehme 5.1 breaks the spotbugs plugin due to internal API usage. Given that these containers are effectively immutable, the callback decoration doesn't buy all that much here.

Options:

1. Do nothing, put the onus on the spotbugs plugin to fix by not using `TaskReportContainer`
2. Put back the original constructor by not decorating for `TaskReportContainer`
3. Put back the original constructor and leave the new one there (i.e. just don't decorate for third party usage)
4. Put back the original constructor and get the decorator via some side mechanism

I think we should do 2 for 5.1, given that we effectively endorse the spotbugs plugin.

**Comment by ldaley** [comment]
@oehme can you give some guidance on whether we do this for 5.1 or not please.

**Comment by ldaley** [comment]
OK, @breskeby is on it.

**Comment by breskeby** [comment]
I fixed this as suggested and added a smoketest in PR at https://github.com/gradle/gradle/pull/8076

**Comment by rpalcolea** [comment]
Thanks for the quick turnaround! I guess this will go on the next 5.1 nightly

**Comment by slavik112211** [comment]
@3flex "Request to make TaskReportContainer public was a moonshot :)"
For an idea how you can reuse TaskReportContainer in an external reporting plugin, see here:
https://github.com/handmadecode/quill/blob/development/src/main/groovy/org/myire/quill/report/MutableReportContainer.java

**Comment by JLLeitschuh** [comment]
Same issue was raised against ktlint-gradle:
https://github.com/JLLeitschuh/ktlint-gradle/issues/349

**Comment by baron1405** [comment]
Over five years since this issue was opened and still no solution for reporting from plugins. Over the years I just took the hit and implemented the ReportContainer interface. While it might not have been intended for clients to implement, I prefer the consistent integration with the built-in plugin report handling to waiting, and waiting, and waiting...
